services/src/main/java/org/keycloak/theme/FreeMarkerUtil.java

```diff
@@ -73,8 +73,8 @@ public String processTemplate(Object data, String templateName, Theme theme) thr
     }
 
     private Template getTemplate(String templateName, Theme theme) throws IOException {
-        Configuration cfg = new Configuration();
-        
+        Configuration cfg = new Configuration(Configuration.VERSION_2_3_26);
```

Comment:
Not sure to be honest as I don't know what the implications really are here. It seems this requires FreeMarker to do additional reflection of the classes, which could be expensive? Also, what would break in existing templates with this?

Reply:
Yes in the sense that it requires FreeMarker to also see default methods, but I'd think this should be done in the same pass where Freemarker retrieves the other getter/setter methods, so this shouldn't have a large overhead (if any).

However, some things may break: Here is an overview: https://freemarker.apache.org/docs/api/freemarker/template/Configuration.html#Configuration-freemarker.template.Version- - since I have no idea what features are used, it's hard for me to judge if this is a problem. I ran some console tests and all seemed fine, so I was confident enough to propose this change.

As I said, the alternative would be to not use default methods but duplicate the methods to all implementing classes.

Reply:
Just an idea, but will it help to use the "wrapper" object, which has the methods getX/setX overriden? I can imagine some change like this:
```
-  freemarkerContextAttributes.put("user", userModel);
+ UserModelDelegate wrapper = new UserModelDelegate(userModel);
+ freemarkerContextAttributes.put("user", wrapper);
```
Can it help?

If not, is it the possibility to have some other Wrapper class similar to for example org.keycloak.forms.login.freemarker.model.ProfileBean ?

If nothing of those helps (or it becomes unecessary complex and ugly code), if you have an opportunity to add the details what exactly the error is and where it happens, it will be helpful. Hopefully we can figure something.

The set of version in the Configuration may work as the last possibility, but will be nice to avoid since we are not 100% sure of all the possible side-effects.

Reply:
I'm sure we'll figure something out!

The problem is actually very simple: I found it first in the `resource-detail.ftl`. The corresponding `AuthorizationBean` contains permissions which contain a `RequesterBean` which contains a `UserModel` (the requester). Now, without this change, the ftl-template cannot be rendered correctly, because `permission.requester.username` is null in this line: https://github.com/keycloak/keycloak/blob/3d5e9760977ac84f627abe28a2df91dd860c13ff/themes/src/main/resources/theme/base/account/resource-detail.ftl#L121 

And the reason it's null is because this refers to the method `getUsername` on the `UserModel`- but it doesn't exist for FTL as default methods don't exist for without this change (this is due to the JavaBean specification which doesn't understand default methods).

Now, clearly that's just one instance of where this fails as the UserModel is also present in other beans. Also, since get/setUsername are not default implemented anymore this particular issue will not arise anymore, but the same problem will be whenever we need get/setEmail or firstName, lastName.

So for me, there would be several solutions to this:

- directly grab the attribute of the user in the FTL templates via something like `permission.requester.attributes["username"][0]` or similar. This is not very nice and very error prone as I don't see a good way to find all instances where this could be a problem
- Add the configuration option as above: least intrusive from a code perspective which is why I proposed this change first, but may have unwanted side-effects. I've never really worked with FTL, so I can't say much about it
- Remove the default methods and just implement them on all subclasses. Since we do essentially the same for `getUsername` it seems those subclasses aren't too many due to the hierarchy of subclasses.
- Add a wrapper object to all places where UserModel is used in the beans. Sounds weird and error prone and results in quite a few changes
- Add an abstract subclass of UserModel and derive all classes that don't override the default methods from this subclass: I don't particularly like to add yet another level to the hierarchy, it's already quite deep
- Make UserModel an abstract class instead of an interface (I haven't tested it though): This should also work, but may not be nice from an architectural standpoint.

I'm probably going to try out the third option. This leads to some code duplication, but maybe this is better for the time being?

Reply:
As I pointed, if there is some easy way to address this like for example using this in the AuthorizationBean constructor: `this.user = new UserModelDelegate(user);`, it will be fine to me.

The solution with adding to subclasses however works for me as well.

Reply:
I ended up adding subclasses - please have a look. I feel that this approach is the safest, as everything looks as before for FTL templates.

All tests already ran beforehand since the only test failure I saw was actually related to the username field and I had already removed the default methods before. I tried using wrapper objects but was left unsatisfied whether I actually solved all problems and if somebody uses custom FTL files, those could easily break if I overlook something. 

I could add a comment to the wrapper class (and the one set of duplicate methods in the infinispan/UserAdapter) to move to default methods once this is no longer a problem in FTL (maybe you will make a switch to a newer version for different reasons).